### ISSUE TITLE
vconsole: if KD_FONT_OP_SET fails (maybe because X11/wayland/some display manager are blocking tty7), debug log and proceed gracefully instead of failing fatally

### ISSUE THREAD

**Issue by eatobin** [systemd/systemd#6689]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> Unsure

### Used distribution

> Unsure

### In case of bug report: Expected behaviour you didn't see

> See below

### In case of bug report: Unexpected behaviour you saw

> "[FAILED] Failed to start Setup Virtual Console."

    eric@linux-epth:~$ sudo systemctl status systemd-vconsole-setup.service
    [sudo] password for root: 
    ● systemd-vconsole-setup.service - Setup Virtual Console
       Loaded: loaded (/usr/lib/systemd/system/systemd-vconsole-setup.service; static; vendor preset: disabled)
       Active: inactive (dead) since Sat 2017-08-26 08:54:42 CDT; 1min 51s ago
         Docs: man:systemd-vconsole-setup.service(8)
           man:vconsole.conf(5)
      Process: 441 ExecStart=/usr/lib/systemd/systemd-vconsole-setup (code=exited, status=0/SUCCESS)
     Main PID: 441 (code=exited, status=0/SUCCESS)

    Aug 26 08:54:42 linux-epth systemd-vconsole-setup[441]: KD_FONT_OP_SET failed, fonts will not be copied to tty7: Invalid argument
    Aug 26 08:54:42 linux-epth systemd[1]: Started Setup Virtual Console.
    eric@linux-epth:~$ sudo systemctl start systemd-vconsole-setup.service
    eric@linux-epth:~$ sudo systemctl status systemd-vconsole-setup.service
    ● systemd-vconsole-setup.service - Setup Virtual Console
       Loaded: loaded (/usr/lib/systemd/system/systemd-vconsole-setup.service; static; vendor preset: disabled)
       Active: inactive (dead) since Sat 2017-08-26 08:57:03 CDT; 4s ago
         Docs: man:systemd-vconsole-setup.service(8)
           man:vconsole.conf(5)
      Process: 2637 ExecStart=/usr/lib/systemd/systemd-vconsole-setup (code=exited, status=0/SUCCESS)
     Main PID: 2637 (code=exited, status=0/SUCCESS)

    Aug 26 08:57:02 linux-epth systemd[1]: Starting Setup Virtual Console...
    Aug 26 08:57:03 linux-epth systemd[1]: Started Setup Virtual Console.
    eric@linux-epth:~$

### In case of bug report: Steps to reproduce the problem

> See above

**Comment by poettering** [comment]
Which version is this?

**Comment by poettering** [comment]
hmm, @msoltyspl any idea about this one?

my educated guess is that DRM drivers or so are loaded/initialized in parallel to the vconsole stuff, and thus things are racy. Given that nobody wants to touch the kernel's VT code these days anymore, it's probably something we just have to accept and handle gracefully, and ignore errors instead of tripping up.

**Comment by eatobin** [comment]
I'm sorry but am not currently on that machine to report a version for you. What console command should I run to obtain the version for you when I can? I will also tell you that the error appears intermittently - so maybe "racy"?

Also reported this error [here](https://bugzilla.opensuse.org/show_bug.cgi?id=1055834) with more detail and commentary from others.

Thank you for your quick response.

**Comment by poettering** [comment]
`systemctl --version` lets you know the version of systemd you are runnin

**Comment by poettering** [comment]
hmm, btw, the fact that this is tty7 is strange. Is it always tty7 that this is reported about? is there somthing special running on tty7?

**Comment by eatobin** [comment]
eric@linux-epth:~$ systemctl --version
    systemd 234
    +PAM -AUDIT +SELINUX -IMA +APPARMOR -SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN2 -IDN default-hierarchy=hybrid
    eric@linux-epth:~$ sudo systemctl status systemd-vconsole-setup.service
    [sudo] password for root: 
    ● systemd-vconsole-setup.service - Setup Virtual Console
       Loaded: loaded (/usr/lib/systemd/system/systemd-vconsole-setup.service; static; vendor preset: disabled)
       Active: inactive (dead) since Tue 2017-08-29 09:26:31 CDT; 4min 51s ago
         Docs: man:systemd-vconsole-setup.service(8)
           man:vconsole.conf(5)
      Process: 426 ExecStart=/usr/lib/systemd/systemd-vconsole-setup (code=exited, status=0/SUCCESS)
     Main PID: 426 (code=exited, status=0/SUCCESS)

    Aug 29 09:26:31 linux-epth systemd-vconsole-setup[426]: KD_FONT_OP_SET failed, fonts will not be copied to tty7: Invalid argu
    Aug 29 09:26:31 linux-epth systemd[1]: Started Setup Virtual Console.
    eric@linux-epth:~$

I don't think there is anything new or special on tty7 - at least that I have added/changed. The behavior is new and I've only applied Tumbleweed updates. Started noticing the changes a few days ago - maybe after [20170822](https://openqa.opensuse.org/factory-package-news/diff/20170822)?

**Comment by eatobin** [comment]
PS - Yes - tty7 is always referenced.

**Comment by poettering** [comment]
And if you switch to tty7 via Alt-F7, what do you see there?

**Comment by eatobin** [comment]
Alt-F7 produces no results, no changes.

**Comment by eatobin** [comment]
Perhaps `dracut` [here](https://openqa.opensuse.org/factory-package-news/diff/20170821)?

**Comment by eatobin** [comment]
One last thought. I did receive this comment from a "Global Moderator" on the openSUSE Forum:

"Sorry, I have absolutely no idea what this could be.

Maybe file a bug report as has been suggested already.

Or ignore it if you don't experience any problems. (and you shouldn't, as this is about being unable to set the text mode font on VT7, but on VT7 runs the graphical interface anyway, maybe that's even the reason for this "error" message)"

Thank you...

**Comment by poettering** [comment]
hu? VT7 runs your graphical interface? that's most likely the cause then. Normally it should run on VT1, and other schemes mean your distro deviates from the suggested defaults...

I guess if X11/wayland is already running on VT7 the font configuration can't work. Hence the error is to be expected, but we probably should proceed ignoring it...

**Comment by eatobin** [comment]
Thank you. But doesn't this strike you as odd? Would there be a change to
VT# as part of an OS update and one that put it at odds with other distros?
Do you know how I might confirm this?

Sent from my rotary phone.

On Aug 29, 2017 11:16, "Lennart Poettering" <notifications@github.com>
wrote:

hu? VT7 runs your graphical interface? that's most likely the cause then.
Normally it should run on VT1, and other schemes mean your distro deviates
from the suggested defaults...

I guess if X11/wayland is already running on VT7 the font configuration
can't work. Hence the error is to be expected, but we probably should
proceed ignoring it...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/systemd/systemd/issues/6689#issuecomment-325715890>,
or mute
the thread

**Comment by eatobin** [comment]
PS - Wouldn't the error be consistent on boot then?

Sent from my rotary phone.

On Aug 29, 2017 11:49, "Eric Tobin" <eatobin@gmail.com> wrote:

Thank you. But doesn't this strike you as odd? Would there be a change to
VT# as part of an OS update and one that put it at odds with other distros?
Do you know how I might confirm this?

Sent from my rotary phone.

On Aug 29, 2017 11:16, "Lennart Poettering" <notifications@github.com>
wrote:

hu? VT7 runs your graphical interface? that's most likely the cause then.
Normally it should run on VT1, and other schemes mean your distro deviates
from the suggested defaults...

I guess if X11/wayland is already running on VT7 the font configuration
can't work. Hence the error is to be expected, but we probably should
proceed ignoring it...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/systemd/systemd/issues/6689#issuecomment-325715890>,
or mute
the thread

**Comment by poettering** [comment]
> Thank you. But doesn't this strike you as odd? Would there be a change to VT# as part of an OS update and one that put it at odds with other distros? Do you know how I might confirm this?

Once upon a time X11 was running on VT7 on many distros and VT1-6 had getty logins statically assigned. With systemd that's different: We only allocate one login session unconditionally, and that's on tty1 and either graphical or tty dependening how you set up your system. And then on tty2-tty6 logind spawns additional gettys dynamically as you switch to them, but only if you actually do that. That way in the usual case we don't run more than necessary. Moreover, when gdm fast-user-switching is used this is also allocated from tty1-tty6 (and beyond) automatically, whatever is currently free. TTY management is thus a lot more dynamic.

Now, some conservative distros live in the past and want their X11 on tty7 and everything fixed and static like it was in 1995. Maybe yours is one of them, and they undo systemd's default logic... But that's really on them...

**Comment by poettering** [comment]
> PS - Wouldn't the error be consistent on boot then?

well, it might simply be a race: vconsole and whatever takes over tty7 start in parallel and whatever wins first is lucky and the other gets an error...

**Comment by eatobin** [comment]
Thank you very much for your time and attention to this. Especially your helpful explanation of VT history (and my distro)! :-) I hope there is a fix - even though the issue seems to be trivial. I will pass on to the openSUSE forum a link to this thread for whatever can be gained there. My best, Eric

**Comment by msoltyspl** [comment]
For the record, this is only a warning message informing that it couldn't handle tty7 (for whatever reasons) - but it continues with remaining consoles. See line 306 in src/vconsole/vconsole-setup.c

vconsole's service doesn't have RemainAfterExit, so it won't be reported as active (which might initially suggest that it bailed out badly after seeing that warning in the logs).

**Comment by eatobin** [comment]
A question please: why do you think this error may have appeared 1.5 years
after installing openSUSE? Bugging me. :-) Thank you.

**Comment by poettering** [comment]
> For the record, this is only a warning message informing that it couldn't handle tty7 (for whatever reasons) - but it continues with remaining consoles. See line 306 in src/vconsole/vconsole-setup.c

Ah, indeed. This looks good then. But something is still wrong there... The original bug report quoted this error message at boot:

> "[FAILED] Failed to start Setup Virtual Console."

Which indicates that the service failed in some way even though "systemctl status" actually shows nothing fatal. I figure we need to do some more digging there.

@eatobin any chance you can boot with "systemd.log_level=debug" on the kernel cmdline, and then attempt to reproduce the issue with the `[FAILED]` line? If you do, please paste the "journalctl -b" output somewhere, so that we can have a closer look why the service fails?

**Comment by poettering** [comment]
> A question please: why do you think this error may have appeared 1.5 years after installing openSUSE? Bugging me. :-) Thank you.

As races usually go: they don't happen always, and some change in some other component might trigger unrelated races previously never seen.

**Comment by msoltyspl** [comment]
> A question please: why do you think this error may have appeared 1.5 years after installing openSUSE? Bugging me. :-) Thank you.

Another thing that changed recently (and might or might not be relevant) is that vconsole now cycles over all possible ttys (in a very careful way - avoiding anything that looks suspicious). If I remember correctly, tty7-tty12 weren't originally touched. Also, vconsole is more careful and verbose if it cannot do something.

**Comment by eatobin** [comment]
I would be happy to boot with "systemd.log_level=debug" on the kernel
cmdline but have no experience with non-standard booting. Is this something
you could easily explain for Tumbleweed or should I Google it?

**Comment by poettering** [comment]
sorry, i have no experience with tumbleweed. The kernel command line is usually editable in your boot loader, i.e. grub or suchlike. Please ask your distro community for help

**Comment by eatobin** [comment]
![20170825_095740](https://user-images.githubusercontent.com/870698/29883798-f74f71dc-8d76-11e7-8370-8dcf711d64f8.jpg)
[journalctl-results.txt](https://github.com/systemd/systemd/files/1264376/journalctl-results.txt)

Okay, I think I got what you've requested and the error did appear on boot after adding the parameter - so I hope it was captured.

The journalctl output was large so it is in the attached text file.

How is this?

**Comment by poettering** [comment]
You need to run "journalctl -b" as roo to get access to system logs, and not just the logs generated by your own user.

**Comment by eatobin** [comment]
[output.txt](https://github.com/systemd/systemd/files/1264447/output.txt)

Better?

**Comment by eatobin** [comment]
[error.txt](https://github.com/systemd/systemd/files/1264465/error.txt)

One more bit of info FYI.

**Comment by poettering** [comment]
Ah, this reveals it. It's not actually vconsole at fault here, but your /usr/bin/loadkeys tool fails with a failure exit code (and we invoke that, and propagate the exit code).

The precise command we invoke is this:

/usr/bin/loadkeys -q -C /dev/tty1 -u us

Which results in a tonload "Invalid argument" messages from loadkeys. Does it work if you invoke it locally from a shell?

It appears to me you should contact your distro for help around this, as it's the kbd package at fault here, not systemd/vconsole

**Comment by eatobin** [comment]
eric@linux-epth:~$ sudo /usr/bin/loadkeys -q -C /dev/tty1 -u us
    [sudo] password for root: 
    eric@linux-epth:~$

"Does it work if you invoke it locally from a shell?"

Looks like... yes?

**Comment by msoltyspl** [comment]
Btw, the errors happen at dracut stage - which would suggest that there might be some frictions between kbd and dracut module responsible for installing relevant stuff after the update (just wild guessing here) - whether they don't copy something, copy it in the wrong place, permissions mismatch or some other missing/unexpected stuff during actual boot or initramfs preparation. Maybe something else is missing that happens to be used by loadkeys but is not related to this particular dracut module itself.

On a side note - unless you need console setup at initramfs stage, you can just blacklist related module and rebuild initramfs image.

If you order dracut to give you debug console before switching root, does this line executed manually work ?

**Comment by alexandru-calinoiu** [comment]
I've hit on the same issue.

**Comment by fbuihuu** [comment]
Some updates: it looks like its due to _plymouth-start.service_ (shipped by plymouth) which has `Wants=systemd-vconsole-setup.service`. This causes the service to be started very early, even before udevd is started from inside the initramfs. At that point loadkeys simply fails as shown previously, not sure why...

Some questions:

 - isn't _systemd-vconsole-setup_ (the binary) supposed to be called **only** from a udev rule only these days ?

 - if so why do we still ship _systemd-vconsole-setup.service_ ?

 - Any idea why _loadkeys_ fails if run before udevd is ?

**Comment by poettering** [comment]
> isn't systemd-vconsole-setup (the binary) supposed to be called only from a udev rule only these days ?

See 8125e8d38e3aa099c7dce8b0161997b8842aebdc

That said, maybe ply pulls it in explicitly in order to ensure that that the screen font is loaded before ply takes over DRM or so. Dunno. @halfline any idea?

> Any idea why loadkeys fails if run before udevd is ?

Maybe your initrd doesn't have the keymaps installed?

**Comment by fbuihuu** [comment]
> Maybe your initrd doesn't have the keymaps installed?

It is installed since the command works if plymouth is disabled.

**Comment by fbuihuu** [comment]
Also the errors are not limited to loadkeys but every attempts to initialize the console:

> systemd-vconsole-setup[198]: KD_FONT_OP_GET failed while trying to get the font metadata: Input/output error
systemd-vconsole-setup[198]: Fonts will not be copied to remaining consoles

**Comment by fbuihuu** [comment]
@poettering couldn't be this related to the graphic driver and when it's exactly loaded by kernel ?

From the logs that someone provided to me, I can see that the nvidia driver is loaded slightly differently in the case systemd-vconsole-setup.service fails...

Could it be the driver that is loaded at the same time vconsole-setup runs ?

**Comment by eatobin** [comment]
I have noticed this too. No or few failures on a machine with nouveau and
repeated errors on my old machine running i915. Also, please remember, this
was not a problem until last month on either computer. Any idea what
changed in TW that began the error?

**Comment by halfline** [comment]
> That said, maybe ply pulls it in explicitly in order to ensure that that the screen font is loaded before ply takes over DRM or so. Dunno. @halfline any idea?

Not sure.  git log has this amazingly well explained message:

```commit 5f3ebd4f6d6c2bf82a9ed48862e86d2663f206c4
Author: Ray Strode <rstrode@redhat.com>
Date:   Tue Mar 26 11:41:14 2013 -0400

    systemd: add Wants: systemd-vconsole-setup.service
    
    Harald says we need it.
```

There's also https://lists.freedesktop.org/archives/systemd-devel/2014-November/024923.html when the systemd-vconsole-setup stuff first got moved to a udev rule.

I guess we can probably just drop it from plymouth-start.service if it's causing problems.

**Comment by fbuihuu** [comment]
Currently I don't think plymouth is the real culprit here although if the dependency on systemd-vconsole-setup.service is useless then it should be dropped.

But it seems that the issue is a timing one: it seems to depend on when systemd-vconsole-setup is executed compare to when the kernel loads the gfx driver.

**Comment by halfline** [comment]
i do have some vague inkling of running across a kernel bug where something that program does (font setup maybe? could be!) doesnt work when plymouth holds drm master. its been years so i might be just inventing it.  you could test the theory by calling plymouth deactivate before it runs and plymouth reactivate after (though it'll have other side effects)

**Comment by fbuihuu** [comment]
Again plymouth isn't the culprit here, at least in my understanding. According to the logs, plymouth hasn't started yet when vconsole-setup start failing.

plymouth just pulls vconsole-setup in (which may be wrong these days) and might trigger vconsole-setup earlier at the time the gfx driver is loaded.

**Comment by aenda** [comment]
Hi - I know this is an old issue, but I'm still getting this warning and wanted to report it. I have plymouth installed on arch linux, and I have early KMS enabled (i915 module enabled in my mkinitcpio.conf). Interestingly enough, when I added the following drop in file for plymouth-start.service:
```
# /etc/systemd/system/plymouth-start.service.d/override.conf
[Unit]
Wants=
Wants=systemd-ask-password-plymouth.path
After=
After=systemd-udev-trigger.service systemd-udevd.service
```
I still get the issue. Any advice?

**Comment by fbuihuu** [comment]
@poettering more information on this issue: It appears that both dracut-cmdline-ask.service and systemd-vconsole-setup.service are started during the early boot and both operates on the console.

dracut-cmdline-ask.service has `StandardInput=tty`, which means that the service will become the controlling process of the terminal and once the service exits the terminal will be in the HUP state.

So if both services are started at the same time and dracut-cmdline-ask.service exits before systemd-vconsole-setup.service (which is likely because this service is a nop if rd.cmdline=ask is not passed) then the terminal will be left in the HUP state and every attempts to configure the terminal by systemd-vconsole.service will fail.

A fix might be to make sure that dracut-cmdline-ask.service is run after systemd-vconsole-setup.service but wouldn't it be more robust to make sure that systemd-vconsole-setup acquire the terminal before it configures it ?

**Comment by mikhailnov** [comment]
Dracut does not include `/etc/vconsole.conf` into initrd by default. openSUSE uses dracut.See https://github.com/dracutdevs/dracut/issues/796

So, /etc/vconsole.conf is empty when systemd-vconsole-setup is run, then this code inside font_load_and_wait() makes systemd-vconsole-setup exit:

if (isempty(font) && isempty(map) && isempty(unimap))
                return 0;

Why does not it jsut run`setfont` without arguements? Maybe it must run setfont despite font being not specified?

**Comment by mikhailnov** [comment]
> if (isempty(font) && isempty(map) && isempty(unimap))
> return 0;
> 
> Why does not it jsut run`setfont` without arguements? Maybe it must run setfont despite font being not specified?

Removed it in https://github.com/systemd/systemd/pull/15479

**Comment by keszybz** [comment]
This should be fixed now. Dracut services mentioned in https://github.com/systemd/systemd/issues/6689#issuecomment-615987458 should be fixed by e019ea738d and graphical services mentioned in https://github.com/systemd/systemd/issues/6689#issue-253668064 should be fixed by a6f3a7eb8a. Please open a new issue if the this still occurs with those patches.
